examples/mlperf/rnnt/train.py

```diff
@@ -0,0 +1,367 @@
+
+    self.params = list(itertools.chain(*[[cell.bias_hh,cell.bias_ih,cell.weights_hh,cell.weights_ih] for cell in self.cells]))
+
+  def do_step(self, x, hc):
+    new_hc = [x]
+    for i, cell in enumerate(self.cells):
+      new_hc.append(cell(new_hc[i][:x.shape[0]], hc[i]))
+    return Tensor.stack(new_hc[1:]).realize()
+
+  def __call__(self, x, hc):
+    @TinyJit
+    def _do_step(x_, hc_):
+      return self.do_step(x_, hc_)
+
+    if hc is None:
+      hc = Tensor.zeros(self.layers, 2 * x.shape[1], self.hidden_size, requires_grad=False)
+
+    output = None
+    for t in range(x.shape[0]):
+      hc = _do_step(x[t] + 1 - 1, hc) # TODO: why do we need to do this?
+      if output is None:
+        output = hc[-1:, :x.shape[1]]
+      else:
+        output = output.cat(hc[-1:, :x.shape[1]], dim=0).realize()
+
+    return output, hc
+
+
+class Joint:
+  def __init__(self, vocab_size, pred_hidden_size, enc_hidden_size, joint_hidden_size, dropout):
+    self.dropout = dropout
+    self.l1 = Linear(pred_hidden_size + enc_hidden_size, joint_hidden_size)
+    self.l2 = Linear(joint_hidden_size, vocab_size)
+    self.params = [self.l1.bias, self.l1.weight, self.l2.bias, self.l2.weight]
+
+  def __call__(self, f, g):
+    (_, T, H), (B, U, H2) = f.shape, g.shape
+    f = f.unsqueeze(2).expand(B, T, U, H)
+    g = g.unsqueeze(1).expand(B, T, U, H2)
+
+    inp = f.cat(g, dim=3)
+    t = self.l1(inp).relu()
+    t = t.dropout(self.dropout)
+    return self.l2(t)
+
+class Encoder:
+  def __init__(self, input_size, hidden_size, pre_layers, post_layers, stack_time_factor, dropout):
+    self.pre_rnn = LSTM(input_size, hidden_size, pre_layers, dropout)
+    self.stack_time = StackTime(stack_time_factor)
+    self.post_rnn = LSTM(stack_time_factor * hidden_size, hidden_size, post_layers, dropout)
+    self.params = [*self.pre_rnn.params, *self.post_rnn.params]
+
+  def __call__(self, x:Tensor, x_lens):
+    x, _ = self.pre_rnn(x, None)
+    x, x_lens = self.stack_time(x, x_lens)
+    x, _ = self.post_rnn(x, None)
+    return x.transpose(0, 1), x_lens
+  
+class StackTime:
+  def __init__(self, factor):
+    self.factor = factor
+
+  def __call__(self, x, x_lens):
+    x = x.pad(((0, (-x.shape[0]) % self.factor), (0, 0), (0, 0)))
+    x = x.reshape(x.shape[0] // self.factor, x.shape[1], x.shape[2] * self.factor)
+    return x, x_lens / self.factor if x_lens is not None else None
+
+class Prediction:
+  def __init__(self, vocab_size, hidden_size, layers, dropout):
+    self.hidden_size = hidden_size
+
+    self.emb = Embedding(vocab_size - 1, hidden_size)
+    self.rnn = LSTM(hidden_size, hidden_size, layers, dropout)
+
+    self.params = [self.emb.weight,*self.rnn.params]
+
+  def __call__(self, x, hc, m):
+    emb = self.emb(x) * m
+    x_, hc = self.rnn(emb.transpose(0, 1), hc)
+    return x_.transpose(0, 1), hc
+
+# %% [markdown]
+# # training
+
+# %%
+def calc_loss(enc,distribution):
```

Comment:
This 100% needs to be done in tinygrad, you can use a custom function for it if the load/store stuff is way too hard to implement.

This breaks the gradient as well no?

Reply:
I decided to calculate the gradient completely manually here since its very fast and my tinygrad implementation was quite slow. A problem here is the gradient get scattered a lot and the AST gets very big. I saw this was a problem in the previous attempts.
Also METAL only supports only float32 that is not enought since the gradient calculation includes quite extreme numbers. Basically for each item in the sequence the activation goes /= 26 and than when in the back pass it gets normalized again. but for the calculation doesnt require a lot of flops so i thought its fine to do this part in python. 
```plt.imshow(distribution_gradient[0].sum(-1)```
should visualizes the gradient I calculated in numpy. This should correspond to Pr in the original RNNT [paper](https://arxiv.org/pdf/1211.3711.pdf) equation (19)
so after I just use the fake _loss to inject gradient into the tinygrad pipeline and let that backprop through the model.

Reply:
I could put it in a Function if this is really required. Would it be okay to leave some computation in numpy though? If no numpy is allowed would need to find a way to make it fast and find a workaround for big numbers in f32. there might be a way to use the logarithm instead of the actual values but not sure since there are a lot of additions going on